### ISSUE TITLE
dotnet restore fails when custom sources (local+url) are specified

### ISSUE THREAD

**Issue by alexeypukhov** [NuGet/Home#7413]
Nuget appends nuget source url to the current work dir and treats it as a local folder instead of url when I pass runtime and two sources: one local path + one url

When I remove runtime it starts to work

--------------
dotnet --info:
.NET Core SDK (reflecting any global.json):
 Version:   2.1.403
 Commit:    04e15494b6

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.17134
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\2.1.403\

Host (useful for support):
  Version: 2.1.5
  Commit:  290303f510

---------------
Repro:
1. dotnet new console
2. dotnet publish -r win-x64 --source ./ --source https://api.nuget.org/v3/index.json

Error:
`C:\Program Files\dotnet\sdk\2.1.403\NuGet.targets(114,5): error : The local source 'C:\Workspace\test\https:\api.nuget.org\v3\index.json' doesn't exist. [C:\Workspace\test\test.csproj]`

[1.log](https://github.com/NuGet/Home/files/2494080/1.log)

**Comment by alexeypukhov** [comment]
@nkolev92 I've checked the issue again and `--packages` flag has nothing to do here.
`-r win-x64` is what triggers the issue. I've updated the issue

**Comment by nkolev92** [comment]
@puhovalexey 

The other options don't really mater.
dotnet.exe forwards the --sources into 1 argument to the NuGet restore target.

It's likely an escaping/ordering problem.

**Comment by alexeypukhov** [comment]
@nkolev92 if that helps I've tried these variants:
 1.a. dotnet publish --source ./ --source https://api.nuget.org/v3/index.json
 1.b. dotnet publish --source ./ --source https://api.nuget.org/v3/index.json -r win-x64
 1.c. dotnet publish --source ./ -r win-x64 --source https://api.nuget.org/v3/index.json
 1.d. dotnet publish  -r win-x64 --source ./ --source https://api.nuget.org/v3/index.json

2.a. dotnet publish --source "./;https://api.nuget.org/v3/index.json"
2.b. dotnet publish --source "./;https://api.nuget.org/v3/index.json" -r win-x64
2.c. dotnet publish  -r win-x64 --source "./;https://api.nuget.org/v3/index.json"

out of which only 1.a. and 2.a. worked. All the other variants gave me the error.
so seems like -r (or some other) parameter makes dotnet.exe treat the sources somehow differently

**Comment by nkolev92** [comment]
yeah, that's interesting. 

Thanks @puhovalexey 
This definitely helps.

**Comment by bradgyton** [comment]
Also seeing this 

.Net Core SDK 2.1.302

dotnet.exe restore SolutionFile.sln --source \\SOMEWHERE\ON\A\NETWORK\ --source https://api.nuget.org/v3/index.json

produces the error

`C:\Program Files\dotnet\sdk\2.1.302\NuGet.targets(114,5): error : The local source 'C:\BuildAgent\work\3f250400dbf075d8\WebApi\https:\api.nuget.org\v3\index.json' doesn't exist.`

**Comment by jDover01** [comment]
Has this been resolved? Facing same issue right now.

**Comment by rrelyea** [comment]
@puhovalexey or @jDover01 - do you see the same problems if you run dotnet restore, instead of dotnet publish?

Not sure if this is related: (from @peterhuene) -https://github.com/dotnet/cli/issues/7791#issuecomment-393756980

**Comment by alexeypukhov** [comment]
@rrelyea I can double check tomo but based on @bradgyton message from Nov 2, 2018 dotnet restore has the same problem:

> Also seeing this
> 
> .Net Core SDK 2.1.302
> 
> dotnet.exe restore SolutionFile.sln --source \SOMEWHERE\ON\A\NETWORK\ --source https://api.nuget.org/v3/index.json
> 
> produces the error
> 
> `C:\Program Files\dotnet\sdk\2.1.302\NuGet.targets(114,5): error : The local source 'C:\BuildAgent\work\3f250400dbf075d8\WebApi\https:\api.nuget.org\v3\index.json' doesn't exist.`

**Comment by alexeypukhov** [comment]
@rrelyea I have double checked and both 'dotnet publish' and 'dotnet restore' give the same error:
`
C:\Program Files\dotnet\sdk\2.2.300\NuGet.targets(121,5): error : The local source 'C:\test\https:\api.nuget.org\v3\index.json' doesn't exist. [C:\test\test.csproj]
`

That's what I did:
1. docker run --rm -ti mcr.microsoft.com/dotnet/core/sdk:2.2-nanoserver-1803
next commands inside a container
2. mkdir test && cd test
3. dotnet new console
4. dotnet publish -r win-x64 --source ./ --source https://api.nuget.org/v3/index.json
5. dotnet restore -r win-x64 --source ./ --source https://api.nuget.org/v3/index.json

**Comment by alexeypukhov** [comment]
@rrelyea interesting thing I repeat the same on linux container `mcr.microsoft.com/dotnet/core/sdk:2.2-stretch` and both 'dotnet publish' and 'dotnet restore' work fine so probably there are some differences between windows and linux

**Comment by jDover01** [comment]
Yes. Whether I used dotnet restore OR dotnet publish on a sln file, I could not supply source arguments. Personally I was trying to restore a custom nuget package for my project. This nuget package lives on the same server that we use for our git repositories. I tried to modify the Nuget.config file for the given user that jenkins uses to log into that box, but that did not work as well. 

My solution was to run the dotnet publish command against the csproj file, and everything worked as expected. (it really shouldn't matter if it is the sln or csproj) Most likely, nuget finally picked up my changes to the Nuget.config file and it started working. (Does the server need to restart if changes are made to Nuget.config file?)

Just tested this again this morning running both restore/publish passing "--source"  argument and I am not receiving any errors from either. =)

**Comment by p3t3rix** [comment]
Any news how to fix this ?
I am having the same problem, same scenario as @bradgyton.  It may be related to the order of the sources because if i put the url one first it resolves just fine (but this is a different behavior)

Those that can't reproduce it anymore: it could be that now the packages are cached, try deleting the local nuget/packages directory.

**Comment by ax-meyer** [comment]
Having the same problem, pretty annoying. Is there a fix now?

**Comment by Alxandr** [comment]
I just spent all of yesterday trying to make this work: https://github.com/YoloDev/YoloDev.Expecto.TestSdk/blob/feat/nbgv/.github/workflows/ci.yml#L116-L123 - because I assumed this error couldn't be in dotnet itself and it only happened on windows. This is incredibly confusing if you run into it and especially when you don't work on windows the error only presents itself on CI making the loop to retry slow.

**Comment by andersstorhaug** [comment]
Also running into this issue. My workaround was to create a separate `NuGet.custom.config` file (named differently in my case so it's not picked up by standard restore) and then to use `--configfile`.

Could not find any other viable workarounds.

I do also have to list the local source before nuget.org, because if I don't, then restore fails when it can't find packages on nuget.org that only exist in the local folder. Maybe this behavior is by design but I thought that I would mention it, because I believe that if I passed nuget.org first via `--sources` then the CLI parsing would actually work, but this behavior prevents me from doing that.

**Comment by TheKnarf** [comment]
>I do also have to list the local source before nuget.org, because if I don't, then restore fails when it can't find packages on nuget.org that only exist in the local folder. Maybe this behavior is by design but I thought that I would mention it, because I believe that if I passed nuget.org first via --sources then the CLI parsing would actually work, but this behavior prevents me from doing that.

Just had a similare problem, had a `nuget.config` that looked like the following:

```
<configuration>
	<config>
		<add key="maxHttpRequestsPerSource" value="4" />
	</config>
	<packageSources>
		<clear />
		<add key="nuget.org" value="https://api.nuget.org/v3/index.json" protocolVersion="3" />
		<add key="DevExpress" value="https://nuget.devexpress.com/secretkey/api" />
	</packageSources>
</configuration>
```

This didn't work at all and the error message was useless. ~~Finally found @andersstorhaug comment and changed up the order of `nuget` and `DevExpress` and suddenly everything works flawlessly:~~

```
<configuration>
	<config>
		<add key="maxHttpRequestsPerSource" value="4" />
	</config>
	<packageSources>
		<clear />
		<add key="DevExpress" value="https://nuget.devexpress.com/secretkey/api" />
		<add key="nuget.org" value="https://api.nuget.org/v3/index.json" protocolVersion="3" />
	</packageSources>
</configuration>
```

**EDIT**

Seems like DevExpress is just flaky.

**Comment by iqapps** [comment]
This problem STILL exist in 2024, was reported first in 2018.

Consider dotnet restore command specifying multiple sources in specific order:

If one specify a folder source first and a https source second, then the https source is formatted internally by dotnet, as if it is another local source with an absolute path. 

Commanding this:

**dotnet restore --source "C:\Code\Packages" --source https://api.nuget.org/v3/index.json**

Throws Error:

**SomeProject.csproj : error NU1301: The local source 'C:\Code\https://api.nuget.org/v3/index.json' doesn't exist. [SomeSoultion.sln]**

This is clearly very wrong.
